aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -447,13 +478,11 @@ Tensor& prod_out(Tensor& result, const Tensor& self, Dimname dim,
 
 Tensor &mean_out_cpu_gpu(Tensor &result, const Tensor &self, IntArrayRef dim,
                  bool keepdim, c10::optional<ScalarType> opt_dtype) {
-  ScalarType scalarType = opt_dtype.has_value() ? opt_dtype.value() : self.scalar_type();
-  TORCH_CHECK(
-      at::isFloatingType(scalarType) || at::isComplexType(scalarType),
-      "Can only calculate the mean of floating types. Got ",
-      toString(scalarType),
-      " instead.");
-  ScalarType dtype = get_dtype(result, self, opt_dtype, true);
+  ScalarType dtype = get_dtype(result, self, opt_dtype);
+  TORCH_CHECK(!at::isIntegralType(dtype, /*includeBool=*/true),
+    "mean requires either a floating output dtype (works on CPU and CUDA) ",
```

Comment:
nit: I know you are trying to be nice with this error message but what's tested should match up with what is checked.  Either remove it from the message and say "doesn't support non-integral" or also check complex and say "must be floating point or complex."  I wouldn't try to record the CPU-vs-CUDA distinction because that will change eventually but the message probably won't.